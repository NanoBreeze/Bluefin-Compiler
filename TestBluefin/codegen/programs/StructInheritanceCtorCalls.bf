/*
If a struct has a parent, its ctor shall call its parent's ctor 
*/

struct Base {
	int a = 234;
};

struct Der extends Base {
	int b;
};

struct DerDer extends Der {
	int b;
};

DerDer derder; 

void foo() {
	DerDer derderTwo;
}