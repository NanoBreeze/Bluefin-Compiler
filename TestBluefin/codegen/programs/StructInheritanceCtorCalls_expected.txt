; ModuleID = 'StructInheritanceCtorCalls.bf'
source_filename = "StructInheritanceCtorCalls.bf"

%DerDer = type { %Der, i32 }
%Der = type { %Base, i32 }
%Base = type { i32 }

@derder = external global %DerDer
@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 0, void ()* @_VarDeclInitializations, i8* null }]

define void @_ctor_Base(%Base* %this) {
entry:
  %memPtr_a = getelementptr %Base, %Base* %this, i32 0, i32 0
  store i32 234, i32* %memPtr_a, align 4
  ret void
}

define void @_ctor_Der(%Der* %this) {
entry:
  %memCast = bitcast %Der* %this to %Base*
  call void @_ctor_Base(%Base* %memCast)
  ret void
}

define void @_ctor_DerDer(%DerDer* %this) {
entry:
  %memCast = bitcast %DerDer* %this to %Der*
  call void @_ctor_Der(%Der* %memCast)
  ret void
}

define internal void @internalGlobalVarDecl_derder() {
entry:
  call void @_ctor_DerDer(%DerDer* @derder)
  ret void
}

define void @foo() {
entry:
  %alloctmp_derderTwo = alloca %DerDer, align 8
  call void @_ctor_DerDer(%DerDer* %alloctmp_derderTwo)
  ret void
}

define internal void @_VarDeclInitializations() {
entry:
  call void @internalGlobalVarDecl_derder()
  ret void
}
